### PR TITLE
feat(NODE-1465): Add mgmt_mac field to deployment.json

### PR DESCRIPTION
**Pull request by andrewbattat**
Similar to https://github.com/dfinity/ic/pull/1199/files

mgmt_mac was a variable we were injecting into config.ini for testing. This change explicitly defines mgmt_mac in deployment.json (where it belongs)

### CODE REVIEW

File: rs/ic_os/utils/src/deployment.rs [hunk 1]
```diff
@@ -314,18 +366,8 @@ mod test {
...
-
-        assert_eq!(DEPLOYMENT_STR_NO_CPU, written_deployment);
-
-        let written_cpu_deployment =
-            serde_json::to_string_pretty::<DeploymentJson>(&QEMU_CPU_DEPLOYMENT_STRUCT).unwrap();
-
-        assert_eq!(QEMU_CPU_DEPLOYMENT_STR, written_cpu_deployment);
-
-        let written_multi_url_deployment =
-            serde_json::to_string_pretty::<DeploymentJson>(&MULTI_URL_STRUCT).unwrap();
```
Comment (andrewbattat) [inline rs/ic_os/utils/src/deployment.rs:327]: I deleted these tests because they were breaking now that the Deployment struct has a new field—because when it would serialize, say, QEMU_CPU_DEPLOYMENT_STRUCT, it would add the mgmt_mac field as null, but the existing QEMU_CPU_DEPLOYMENT_STR wouldn't have that field.

I figured it was okay to delete these tests because we really only care about deserializing the json. We never serialize it back to json, so no need to test it.

File: ic-os/setupos/data/deployment.json.template [hunk 2]
```diff
@@ -1,6 +1,7 @@
 {
   "deployment": {
-    "name": "mainnet"
+    "name": "mainnet",
+    "mgmt_mac": null
```
Comment (andrewbattat) [inline ic-os/setupos/data/deployment.json.template:4]: I added the mgmt_mac field to deployment because doing so was easier for the parsing backwards compatibility than creating a new "network" containing key, but I can add a network key if that feels more proper

File: rs/ic_os/network/src/mac_address.rs [hunk 3]
```diff
@@ -121,19 +121,17 @@ fn generate_mac_address_internal(
...
-        None => {
-            let ipmitool_output = get_command_stdout("ipmitool", ["lan", "print"])?;
-            get_mac_address_from_ipmitool_output(&ipmitool_output)?
-        }
-    };
-    generate_mac_address_internal(&bmc_mac, deployment_name, node_type, '6')
+    let mgmt_mac = if let Some(mgmt_mac) = mgmt_mac {
+        let mgmt_mac = FormattedMacAddress::try_from(mgmt_mac)?;
+        eprintln!("Using mgmt_mac address found in config: {}", mgmt_mac.get());
+        Ok(mgmt_mac.clone())
```
Comment (Bownairo) [inline rs/ic_os/network/src/mac_address.rs:129]: Do we need to clone here, now that we're building this struct inside?

File: rs/ic_os/hostos_tool/src/main.rs [hunk 4]
```diff
@@ -73,8 +74,14 @@ pub fn main() -> Result<()> {
                 }
             };
 
+            let mgmt_mac: Option<&str> = match &deployment {
```
Comment (Bownairo) [inline rs/ic_os/hostos_tool/src/main.rs:77]: Maybe, could we do the parsing here so that we have the typed `FormattedMacAddress` earlier?

File: rs/ic_os/utils/src/deployment.rs [hunk 5]
```diff
@@ -80,21 +81,42 @@ mod test {
     use once_cell::sync::Lazy;
     use serde_json::{json, Value};
 
+    static DEPLOYMENT_VALUE: Lazy<Value> = Lazy::new(|| {
+        json!({
+              "deployment": {
+                "name": "mainnet",
+                "mgmt_mac": null
```
Comment (Bownairo) [inline rs/ic_os/utils/src/deployment.rs:88]: Since this is optional, does we need to specify this here?

File: rs/ic_os/utils/src/deployment.rs (hunk 1 above)
Comment (Bownairo) [reply rs/ic_os/utils/src/deployment.rs:327]: > I figured it was okay to delete these tests because we really only care about deserializing the json. We never serialize the deployment struct back to json, so no need to test it.

Do you mean that there's a lot of redundancy in tests because we are still keeping the one serialization test?

In general, I don't think it hurts to have some extra tests, especially since we may soon build on it with the rest of the config work.

If the only issue is that we're seeing the value serialize as `null`, I think we should change this behavior in the serializer instead. `#[serde(skip_serializing_if = "Option::is_none")]` is a pretty common pattern ([source](https://serde.rs/field-attrs.html#skip_serializing_if)).

File: ic-os/setupos/data/deployment.json.template (hunk 2 above)
Comment (Bownairo) [reply ic-os/setupos/data/deployment.json.template:4]: I think this should be fine since the field is only used in tests, but even changes like this can bite us with backwards compatibility. Once we move the field, our repo tools won't be able to modify old versions of images. We're not often testing too far back in time, though, and it should be easy enough to modify directly when needed.

File: rs/ic_os/utils/src/deployment.rs (hunk 5 above)
Comment (Bownairo) [reply rs/ic_os/utils/src/deployment.rs:88]: (Following other comments, we might be able to remove this.)

File: rs/ic_os/utils/src/deployment.rs (hunk 1 above)
Comment (andrewbattat) [reply rs/ic_os/utils/src/deployment.rs:327]: The way I understand it is the read_deployment test is the only one that really matters. read_deployment reads in the DEPLOYMENT_STR (deployment.json) and converts it to the DEPLOYMENT_STRUCT. Example:
```
let parsed_deployment: DeploymentJson = { serde_json::from_str(DEPLOYMENT_STR).unwrap() };

        assert_eq!(*DEPLOYMENT_STRUCT, parsed_deployment);
```

This is exactly what we do in production.

The write_deployment test converts the DEPLOYMENT_STRUCT back into a DEPLOYMENT_STR (deployment.json), but we never do this in production. We only deserialize, never serialize.

So when #[serde(skip_serializing_if = "Option::is_none")] is added and we serialize DEPLOYMENT_STRUCT in our test, it's actually not DEPLOYMENT_STR because DEPLOYMENT_STR has been updated to include mgmt_mac. Instead,  DEPLOYMENT_STRUCT serializes to _DEPLOYMENT_STR_NO_MGMT_MAC_. 

So we could make these changes to get the test to work. But like I said, read_deployment is the only test that really matters because we never serialize the deployment struct and I don't think we ever will. For this reason, I figured we could delete the test, but I left the single write_deployment assertion as a bit of a sanity check. Does that all make sense? WDYT?

File: rs/ic_os/hostos_tool/src/main.rs (hunk 4 above)
Comment (andrewbattat) [reply rs/ic_os/hostos_tool/src/main.rs:77]: I started on this but then the changes started to really balloon, so I made a separate branch that I can pick up after this PR merges: https://github.com/dfinity/ic/pull/1397/files (see latest commit)

WDYT?

File: rs/ic_os/utils/src/deployment.rs (hunk 1 above)
Comment (Bownairo) [reply rs/ic_os/utils/src/deployment.rs:327]: > This is exactly what we do in production.

I think it's OK to keep the "main" test in line with the current repo's config, but we also have all the state currently deployed to worry about. Checking that new code doesn't break with old values is exactly what these tests help us catch.

> But like I said, read_deployment is the only test that really matters because we never serialize the deployment struct and I don't think we ever will.

IMO more generally, tests like this give us the chance to check boundaries of code, even for cases that may not yet be fully exercised in reality. Though, I'm pretty sure we _are_ using serialization in [`setupos-inject-configuration`](https://github.com/dfinity/ic/blob/e93ea2e7492b6b66dcba3ab76fad1a168b234097/rs/ic_os/setupos-inject-configuration/src/main.rs#L253).

On `skip_serializing_if `, being that
```
{
    "key": "value",
    "key2": null
}
```
and
```
{
    "key": "value"
}
```
are equivalent as far as we're concerned, I tend to prefer the second, but really either one should be OK.

Comment (andrewbattat) [reply rs/ic_os/utils/src/deployment.rs]: Sounds good! I updated the test!

**Review by Bownairo — APPROVED** [review]

File: rs/ic_os/hostos_tool/src/main.rs [hunk 6]
```diff
@@ -73,8 +74,14 @@ pub fn main() -> Result<()> {
...
+                Ok(deployment) => deployment.deployment.mgmt_mac.as_deref(),
+                Err(_) => None,
```
Comment (frankdavid) [inline rs/ic_os/hostos_tool/src/main.rs:79]: When can this be an error? Could we fail or at least log the error?

Comment (andrewbattat) [reply rs/ic_os/hostos_tool/src/main.rs:79]: See discussion here: https://github.com/dfinity/ic/pull/1190/files#r1752744231
